internal/engine/fswatch/gitmanager.go

```diff
@@ -0,0 +1,138 @@
+package fswatch
+
+import (
+	"context"
+	"io/ioutil"
+	"path/filepath"
+	"time"
+
+	"github.com/tilt-dev/tilt/internal/store"
+	"github.com/tilt-dev/tilt/internal/watch"
+	"github.com/tilt-dev/tilt/pkg/logger"
+	"github.com/tilt-dev/tilt/pkg/model"
+)
+
+type repoNotifyCancel struct {
+	repo   model.LocalGitRepo
+	notify watch.Notify
+	cancel func()
+}
+
+// Watches git for branch switches and fires events.
+type GitManager struct {
+	fsWatcherMaker FsWatcherMaker
+	repoWatches    map[model.LocalGitRepo]repoNotifyCancel
+}
+
+func NewGitManager(fsWatcherMaker FsWatcherMaker) *GitManager {
+	return &GitManager{
+		fsWatcherMaker: fsWatcherMaker,
+		repoWatches:    make(map[model.LocalGitRepo]repoNotifyCancel),
+	}
+}
+
+func (m *GitManager) diff(ctx context.Context, st store.RStore) (setup, teardown []model.LocalGitRepo) {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	if !state.EngineMode.WatchesFiles() {
+		return nil, nil
+	}
+
+	watchable := WatchableTargetsForManifests(state.Manifests())
+	reposToProcess := make(map[model.LocalGitRepo]bool)
+	for _, w := range watchable {
+		for _, repo := range w.LocalRepos() {
+			reposToProcess[repo] = true
+		}
+	}
+
+	for repo := range reposToProcess {
+		if _, ok := m.repoWatches[repo]; !ok {
+			setup = append(setup, repo)
+		}
+	}
+
+	for repo := range m.repoWatches {
+		if _, ok := reposToProcess[repo]; !ok {
+			teardown = append(teardown, repo)
+		}
+	}
+
+	return setup, teardown
+}
+
+func (m *GitManager) OnChange(ctx context.Context, st store.RStore) {
+	setup, teardown := m.diff(ctx, st)
+
+	for _, repo := range teardown {
+		p, ok := m.repoWatches[repo]
+		if !ok || p.notify == nil {
+			continue
+		}
+		_ = p.notify.Close()
+		p.cancel()
+		delete(m.repoWatches, repo)
+	}
+
+	for _, repo := range setup {
+		m.repoWatches[repo] = repoNotifyCancel{repo: repo}
+
+		watcher, err := m.fsWatcherMaker(
+			[]string{gitHeadPath(repo)},
```

Comment:
I think this merits a comment about what's going on ("a filewatcher watching for changes to the file that sets the git head. Changes to this file represent..." etc.)